miden-crypto/src/word.rs

```diff
@@ -0,0 +1,614 @@
+
+/// A unit of data consisting of 4 field elements.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[cfg_attr(feature = "serde", serde(into = "String", try_from = "&str"))]
+pub struct Word([Felt; WORD_SIZE_FELT]);
+
+impl Word {
+    /// The serialized size of the word in bytes.
+    pub const SERIALIZED_SIZE: usize = WORD_SIZE_BYTES;
+
+    /// Creates a new [Word] from the given field elements.
+    pub const fn new(value: [Felt; WORD_SIZE_FELT]) -> Self {
+        Self(value)
+    }
+
+    /// Returns the word as a slice of field elements.
+    pub fn as_elements(&self) -> &[Felt] {
+        self.as_ref()
+    }
+
+    /// Returns the word as a byte array.
+    pub fn as_bytes(&self) -> [u8; WORD_SIZE_BYTES] {
+        <Self as Digest>::as_bytes(self)
+    }
+
+    /// Returns an iterator over the elements of multiple words.
+    pub(crate) fn words_as_elements_iter<'a, I>(words: I) -> impl Iterator<Item = &'a Felt>
+    where
+        I: Iterator<Item = &'a Self>,
+    {
+        words.flat_map(|d| d.0.iter())
+    }
+
+    /// Returns all elements of multiple words as a slice.
+    pub(crate) fn words_as_elements(words: &[Self]) -> &[Felt] {
+        let p = words.as_ptr();
+        let len = words.len() * WORD_SIZE_FELT;
+        unsafe { slice::from_raw_parts(p as *const Felt, len) }
+    }
+
+    /// Returns hexadecimal representation of this word prefixed with `0x`.
+    pub fn to_hex(&self) -> String {
+        bytes_to_hex_string(self.as_bytes())
+    }
+}
+
+impl Hash for Word {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(&self.as_bytes());
+    }
+}
+
+impl Digest for Word {
+    fn as_bytes(&self) -> [u8; WORD_SIZE_BYTES] {
+        let mut result = [0; WORD_SIZE_BYTES];
+
+        result[..8].copy_from_slice(&self.0[0].as_int().to_le_bytes());
+        result[8..16].copy_from_slice(&self.0[1].as_int().to_le_bytes());
+        result[16..24].copy_from_slice(&self.0[2].as_int().to_le_bytes());
+        result[24..].copy_from_slice(&self.0[3].as_int().to_le_bytes());
+
+        result
+    }
+}
+
+impl Deref for Word {
+    type Target = [Felt; WORD_SIZE_FELT];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Ord for Word {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Compare the inner u64 of both elements.
+        //
+        // It will iterate the elements and will return the first computation different than
+        // `Equal`. Otherwise, the ordering is equal.
+        //
+        // Finally, we use `Felt::inner` instead of `Felt::as_int` so we avoid performing a
+        // montgomery reduction for every limb. That is safe because every inner element of the
+        // word is guaranteed to be in its canonical form (that is, `x in [0,p)`).
+        self.0.iter().map(Felt::inner).zip(other.0.iter().map(Felt::inner)).fold(
```

Comment:
It may be worth mentioning that the because we don't perform the Montgomery reduction, the ordering is not "lexicographical". That is, looking at the bytestring of two words we wouldn't be able to tell which one is "greater" and which one is "smaller".

Reply:
Updated